Feedback from PRB on V2.2

PaulRBerg opened the pull request:
I spent a few days reviewing the `staging` branch - good job @sablier-labs/solidity!

Closes #870, #871, #878, #879, and #880.

smol-ninja commented on test/integration/shared/lockup-linear/createWithTimestamps.t.sol:
```diff
@@ -26,7 +26,15 @@ abstract contract CreateWithTimestamps_Integration_Shared_Test is LockupLinear_I
     }
 
-    modifier whenStartTimeNotGreaterThanCliffTime() {
+    modifier whenCliffTimeZero() {
```
Wouldn't it be easier to follow modifiers if we order them alphabetically?

PaulRBerg replied:
It would be easier! But if we do this, we should do it in all shared test files containing modifiers. I opened an issue to track this idea: https://github.com/sablier-labs/v2-core/issues/884

smol-ninja commented on test/mocks/erc20/ERC20Mock.sol:
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-License-Identifier: UNLICENSED
```
Why are we using `UNLICENSED` for mocks?

PaulRBerg replied:
It doesn't really matter what we're using here as long it's an open-source license. I reverted the changes to use GPL v3 again.

FWIW, my rationale was that nobody will install Sablier as a dependency just to get access to these mocks.

smol-ninja reviewed (changes requested):
Fantastic review 👏. Those are really good changes. I like using the term `distribution` instead of `streaming`. 

I've added two new commits (https://github.com/sablier-labs/v2-core/pull/883/commits/e88ef6ed6b9080e6a3777c6ded6ab3ebd50018b2 and https://github.com/sablier-labs/v2-core/pull/883/commits/bda5c856a292827950ce65adb35cc401fba637f3) to address some minor changes.

andreivladbrg commented:
Thanks for the review PRB!

Haven't finished reviewing yet, but I want to point out one thing to get things move faster, I guess this was not intentional:

https://github.com/sablier-labs/v2-core/blob/bda5c856a292827950ce65adb35cc401fba637f3/src/SablierV2LockupTranched.sol#L47-L49

`cliffs` mapping on Tranched contract

Doing some changes locally, so I will push commit once I finish.

PaulRBerg commented on src/SablierV2LockupDynamic.sol:
```diff
@@ -342,24 +339,24 @@ contract SablierV2LockupDynamic is
             stream.endTime = params.segments[segmentCount - 1].timestamp;
 
-            // Effects: store the segments. Since Solidity lacks a syntax for copying arrays directly from
+            // Effect: store the tranches. Since Solidity lacks a syntax for copying arrays of structs directly from
```
Of course it should say that. Good catch.

PaulRBerg commented:
Thanks for the counter-review, guys.

> Can you confirm, please, that `cliffs` mapping should not be `Tranched` contract?

Ofc not. I have accidentally copy-pasted it there. Good catch.

smol-ninja commented on test/utils/Precompiles.t.sol:
```diff
@@ -25,7 +25,7 @@ contract Precompiles_Test is Base_Test {
         address actualLockupDynamic = address(precompiles.deployLockupDynamic(users.admin, nftDescriptor));
         address expectedLockupDynamic =
-            address(deployOptimizedLockupDynamic(users.admin, nftDescriptor, defaults.MAX_COUNT()));
+            address(deployOptimizedLockupDynamic(users.admin, nftDescriptor, defaults.MAX_SEGMENT_COUNT()));
```
Precompile tests are failing because this is not same as the values we are using in `Precompiles.sol`.

smol-ninja replied:
I think we should keep it 500 until https://github.com/sablier-labs/v2-core/issues/872 is resolved.

PaulRBerg replied:
Good catch and OK, I will revert to 500.

andreivladbrg reviewed (changes requested):
Great findings and suggestions PRB. Short form for `_create` is better, capitalizing the ID improves readability and also like more and prefer the naming of the functions in the Helpers library now. I remember using the same terminology when the checks were [initially moved to a separate library](https://github.com/sablier-labs/v2-core/pull/154/).

PaulRBerg commented on src/interfaces/ISablierV2LockupLinear.sol:
```diff
@@ -7,20 +7,20 @@ import { Lockup, LockupLinear } from "../types/DataTypes.sol";
 
 /// @title ISablierV2LockupLinear
-/// @notice Creates and manages Lockup streams with linear streaming functions.
+/// @notice Creates and manages Lockup streams with linear distribution functions.
```
It doesn't really matter. We could say "with a linear distribution function".

If you have a strong preference, feel free to make the change.

PaulRBerg replied:
I applied the change myself.

PaulRBerg commented on src/SablierV2LockupTranched.sol:
```diff
@@ -44,7 +44,8 @@ contract SablierV2LockupTranched is
 
-    /// @dev Stream tranches mapped by stream ids.
+    /// @dev Stream tranches mapped by stream IDs. This complements the `_streams` mapping in {SablierV2Lockup}.
+    mapping(uint256 id => uint40 cliff) internal _cliffs;
```
yep, removed

PaulRBerg commented on script/GenerateSVG.s.sol:
```diff
@@ -39,10 +39,10 @@ contract GenerateSVG is BaseScript, SablierV2NFTDescriptor {
                 progressNumerical: progress,
-                status: status,
-                streamingModel: "Lockup Linear"
+                status: status
```
yep

PaulRBerg commented on src/SablierV2LockupDynamic.sol:
```diff
@@ -168,26 +168,26 @@ contract SablierV2LockupDynamic is
     /// - $exp$ is the current segment exponent.
     /// - $csa$ is the current segment amount.
-    /// - $\Sigma(esa)$ is the sum of all elapsed segments' amounts.
+    /// - $\Sigma(esa)$ is the sum of all vested segments' amounts.
```
Yes, it is. An amount does not "elapse" - only time elapses.

"vested" need not refer specifically to token or stock vesting - it's a general-purpose English word. From [Merriam Webster](https://www.merriam-webster.com/dictionary/vest):

> to place or give into the possession or discretion of some person or authority

andreivladbrg replied:
ok, got it, makes sense

PaulRBerg commented on src/SablierV2LockupDynamic.sol:
```diff
@@ -245,20 +245,20 @@ contract SablierV2LockupDynamic is
 
-            // Divide the elapsed segment time by the total duration of the segment.
-            SD59x18 elapsedSegmentTimePercentage = elapsedSegmentTime.div(totalSegmentTime);
+            // Divide the vested segment's time by the total duration of the segment.
```
There's no inconsistency in those comments, but you're right that we should rename `elapsedSegmentTime` to `elapsedTime` and `elapsedSegmentTimePercentage` to `elapsedTimePercentage`.

The rationale is:

- Time _elapses_
- But amounts and segments do not elapse, they _vest_

Combining the logic in a coherent whole:

_Elapsing time vests segments and their amounts._

andreivladbrg replied:
understood now

PaulRBerg commented on src/SablierV2LockupTranched.sol:
```diff
@@ -194,16 +195,16 @@ contract SablierV2LockupTranched is
+        for (uint256 i = 1; i < tranches.length; ++i) {
+            // If a tranche's timestamp is equal to the current time, it is considered vested.
+            if (tranches[i].timestamp > currentTime) {
+                break;
```
Oh, shoot. No reason - it was a mistake.

Correcting now.

PaulRBerg commented:
Thanks for the kind words, @andreivladbrg.

I have addressed most of your feedback with my latest commit.

Are we good to merge now?

The precompile tests are failing, but that's because of https://github.com/sablier-labs/v2-core/issues/886.

smol-ninja commented on src/SablierV2LockupTranched.sol:
```diff
@@ -199,7 +199,7 @@ contract SablierV2LockupTranched is
         uint128 streamedAmount = tranches[0].amount;
         for (uint256 i = 1; i < tranches.length; ++i) {
-            // If a tranche's timestamp is equal to the current time, it is considered vested.
+            // If a tranche's timestamp is less than or equal to the current time, it has vested.
```
```suggestion
            // If a tranche's timestamp is less than or equal to the block time, it has vested.
```

We should also rename this in the comment since we have renamed `currentTime` to `blockTimestamp`.

PaulRBerg replied:
Good catch. I will fix now.

PaulRBerg commented:
Ok guys, I reviewed everything.

Whenever you're ready, we can merge this.

smol-ninja commented:
@andreivladbrg I see you have replaced ERC721 mock with forge mock. Can we also do the same with `ERC20Mock`?

andreivladbrg commented on slither.config.json:
```diff
@@ -1,9 +1,10 @@
   "solc_remaps": [
     "@openzeppelin/contracts/=node_modules/@openzeppelin/contracts/",
     "@prb/math/=node_modules/@prb-math/",
+    "@sphinx-labs/contracts/=node_modules/@sphinx-labs/contracts/contracts/foundry",
```
shouldn't we add "/" at the end?

PaulRBerg replied:
I guess? I'll add it.

PaulRBerg replied:
maybe this was the problem with Sphinx all along? :D

cc @smol-ninja

PaulRBerg commented:
I've also bumped PRBMath (see context here: https://github.com/PaulRBerg/prb-math/pull/225).
